packages/frontend/src/models/CanonicalBridge.ts

```diff
@@ -0,0 +1,341 @@
+  | L1XDaiPoaBridge
+  | L1XDaiWETHOmnibridgeRouter
+
+class CanonicalBridge extends Base {
+  public chain: Chain
+  public tokenSymbol: TokenSymbol
+  address: string
+  l2TokenAddress: string
+
+  constructor(network: string, signer: Signer, token: TToken, chain: TChain) {
+    super(network, signer)
+    if (!token) {
+      throw new Error('token symbol is required')
+    }
+    this.getSignerOrProvider(Chain.Ethereum, signer).then(s => {
+      this.signer = s
+    })
+    this.chain = this.toChainModel(chain)
+    this.tokenSymbol = this.toTokenModel(token).symbol
+    this.address = this.getL1CanonicalBridgeAddress(this.tokenSymbol, this.chain)
+    this.l2TokenAddress = this.getL2CanonicalTokenAddress(this.tokenSymbol, this.chain)
+  }
+
+  public connect(signer: Signer) {
+    return new CanonicalBridge(this.network, signer, this.tokenSymbol, this.chain)
+  }
+
+  public getL1CanonicalAllowance(): Promise<BigNumber> {
+    const l1CanonicalToken = this.getL1Token().connect(this.signer)
+    const spender = this.getDepositApprovalAddress()
+    if (!spender) {
+      throw new Error(`token "${this.tokenSymbol}" on chain "${this.chain.slug}" is unsupported`)
+    }
+    return l1CanonicalToken.allowance(spender)
+  }
+
+  public getDepositApprovalAddress(): string {
+    if (this.chain.equals(Chain.Polygon) && this.tokenSymbol !== Token.MATIC) {
+      return this.getL1PosErc20PredicateAddress(this.tokenSymbol, this.chain)
+    }
+
+    if (this.chain.equals(Chain.Arbitrum) && this.tokenSymbol !== Token.ETH) {
+      return this.getL1ArbitrumGateway(this.tokenSymbol)
+    }
+
+    return this.address
+  }
+
+  public async estimateApproveTx(amount: BigNumberish) {
+    const l1CanonicalToken = this.getL1Token().connect(this.signer)
+    const spender = this.getDepositApprovalAddress()
+    const populatedTx = await l1CanonicalToken.populateApproveTx(spender, amount)
+    return this.signer.estimateGas(populatedTx)
+  }
+
+  public async approve(amount: BigNumberish) {
+    amount = amount.toString()
+    const l1CanonicalToken = this.getL1Token().connect(this.signer)
+
+    const spender = this.getDepositApprovalAddress()
+    if (!spender) {
+      throw new Error(`token "${this.tokenSymbol}" on chain "${this.chain.slug}" is unsupported`)
+    }
+    return l1CanonicalToken.approve(spender, amount)
+  }
+
+  public async estimateDepositTx(amount: BigNumberish) {
+    const populatedTx = await this.populateDepositTx(amount)
+    return this.signer.estimateGas(populatedTx)
+  }
+
+  public async deposit(amount: BigNumberish, customRecipient?: string) {
+    if (this.chain.equals(Chain.Arbitrum) && this.tokenSymbol !== CanonicalToken.ETH) {
+      return this.populateDepositTx(amount, customRecipient)
+    }
+    const populatedTx = await this.populateDepositTx(amount)
+    return this.signer.sendTransaction(populatedTx)
+  }
+
+  public async populateDepositTx(amount: BigNumberish, customRecipient?: string): Promise<any> {
```

Comment:
suggestion: instead of multiple if statements checking chain and token, using polymorphism would be cleaner ([example](https://github.com/hop-protocol/hop/blob/develop/packages/sdk/src/watchers/Watcher.ts))